File: SPOJ-PT07X.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

const int N = 1e5 + 5, M = 1e3 + 5, MOD = 32768, HV = 151, INF = 0x3f3f3f3f;
bool vis[N];
vector<int> tre[N];

void run()
{
    int n;
    scanf("%d", &n);
    if (n == 1)
    {
        printf("1\n");
        return;
    }
    for (int i = 0, u, v; i < n - 1; i++)
    {
        scanf("%d%d", &u, &v);
        tre[u].push_back(v);
        tre[v].push_back(u);
    }
    queue<int> q;
    for (int i = 1; i <= n; i++)
    {
        if (tre[i].size() == 1)
        {
            q.push(i);
            vis[i] = true;
        }
    }
    int ans = 0, p = 0;
    while (q.size())
    {
        ans += q.size() * p;
        queue<int> tem;
        while (q.size())
        {
            int f = q.front();
            q.pop();
            for (auto &&i : tre[f])
            {
                if (vis[i])
                {
                    continue;
                }
                tem.push(i);
                vis[i] = true;
            }
        }
        swap(q, tem);
        p ^= 1;
    }
    printf("%d\n", ans);
}

int main()
{
    // freopen("_input.txt", "r", stdin);
    // freopen("_output.txt", "w", stdout);
    int t = 1;
    // scanf("%d", &t);
    while (t--)
    {
        run();
    }
    return 0;
}

File: SPOJ-PT07X.py
from functools import lru_cache
import sys


def main():
    n = int(input())
    tre = dict()
    for i in range(n - 1):
        u, v = map(int, input().split())
        if u not in tre:
            tre[u] = set()
        if v not in tre:
            tre[v] = set()
        tre[u].add(v)
        tre[v].add(u)

    # @lru_cache
    def solve(n, p):
        ret = (1, -1)
        l = list()
        for i in tre[n]:
            if i == p:
                continue
            l.append(solve(i, n, p ^ 1))
        mn = float("inf")
        for i in l:
            mn = min(mn, i[])
        return ret

    if n == 1:
        print(1)
    else:
        ans = solve(1, 0, 0)
        assert(min(ans, n - ans) == min(min(solve(i+1, 0, 0), solve(i+1, 0, 1)) for i in range(n)))


if __name__ == "__main__":
    with open("_input.txt", "r") as fin, open("_output.txt", "w") as fout:
        sys.stdin = fin
        sys.stdout = fout
        main()
    # main()
